### ISSUE TITLE
Requests have no timeout

### ISSUE THREAD

**Issue by lucymhdavies** [devops-rob/terraform-provider-terracurl#45]
### Terraform Version

TF: 1.3.9

```
- Installed devops-rob/terracurl v1.0.2 (self-signed, key ID C04C4BF84C9276F1)
```

### Affected Resource(s)
I've seen the issue on the resource, but I expect the datasource would have the same problem

### Terraform Configuration Files
```hcl
locals {
  webserver_url = "http://${aws_instance.web.public_ip}"
}

resource "terracurl_request" "test" {
  name   = "smoke test webserver"
  url    = local.webserver_url
  method = "GET"

  response_codes = [
    200
  ]

  // Retry for up to 60s
  max_retry      = 
  retry_interval = 15

  lifecycle {
    // Any change to the EC2 instance, do another TerraCurl check
    replace_triggered_by = [aws_instance.web]

    // And check that we have an HTTP 200
    postcondition {
      condition     = self.status_code == "200"
      error_message = "Webserver did not respond with an HTTP 200"
    }
  }
}
```

To reproduce the issue, I updated the security group on my EC2 instance to remove inbound port 80.

### Expected Behavior
Request should timeout, and TF should retry

### Actual Behavior
Request does not timeout

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids

Looks like it's a relatively simple addition:
https://github.com/devops-rob/terraform-provider-terracurl/blob/main/internal/provider/resource_curl.go#LL328C54-L328C58

`context.TODO()` replace with `context.WithTimeout()`
https://pkg.go.dev/context#WithTimeout

+ relevant parameters to the resource/datasource to configure that with sensible defaults

**Comment by lucymhdavies** [comment]
I'll give https://github.com/devops-rob/terraform-provider-terracurl/releases/tag/v1.2.0-rc1 a try later and see how it goes :)

changes on this branch:
https://github.com/devops-rob/terraform-provider-terracurl/commits/timeout-fix

looking at docs... default timeout value of 10 should be sufficient to validate that it works, and the actual change to the provider looks sound:
https://github.com/devops-rob/terraform-provider-terracurl/commit/870ff9e0f33906d0a7589e8f709d22d8001c218b

**Comment by lucymhdavies** [comment]
```
terraform {
  required_providers {
    terracurl = {
      source  = "terraform.local/local/terracurl"
      version = "1.2.0-rc1"
    }
  }
}
```

```
✔ strawb@HashiBook ~/.terraform.d/plugins/terraform.local/local/terracurl/1.2.0-rc1/darwin_amd64
$ ls
terraform-provider-terracurl_1.2.0-rc1_darwin_amd64.zip terraform-provider-terracurl_v1.2.0-rc1
```

(in case I ever forget how to do local providers again)




and trying with a simple happy path first...


```
resource "terracurl_request" "test" {
  name   = "smoke test webserver"
  url    = module.webserver.public_url
  method = "GET"

  response_codes = [
    200
  ]

  // 4 retries, with default 10 second timeout, with 15 seconds between...
  // = 4 * (10 + 15) - 15 = 4 * 25 - 15 = 85, i.e. 1m25s
  max_retry      = 4
  retry_interval = 15
  timeout = 10
}
```


```
terracurl_request.test: Still creating... [10s elapsed]
module.webserver.azurerm_linux_virtual_machine.example: Still creating... [10s elapsed]
terracurl_request.test: Still creating... [20s elapsed]
module.webserver.azurerm_linux_virtual_machine.example: Still creating... [20s elapsed]
terracurl_request.test: Still creating... [30s elapsed]
module.webserver.azurerm_linux_virtual_machine.example: Still creating... [30s elapsed]
terracurl_request.test: Still creating... [40s elapsed]
module.webserver.azurerm_linux_virtual_machine.example: Still creating... [40s elapsed]
terracurl_request.test: Creation complete after 48s [id=smoke test webserver]
```

so that's looking good...


now for a failing test, i'm removing the network security group rule that allows me to access my instance between running a plan and running an apply (to trick terraform into not just re-creating the rule)



```
terracurl_request.test: Creating...
terracurl_request.test: Still creating... [10s elapsed]
terracurl_request.test: Still creating... [20s elapsed]
terracurl_request.test: Still creating... [30s elapsed]
terracurl_request.test: Still creating... [40s elapsed]
terracurl_request.test: Still creating... [50s elapsed]
terracurl_request.test: Still creating... [1m0s elapsed]
terracurl_request.test: Still creating... [1m10s elapsed]
terracurl_request.test: Still creating... [1m20s elapsed]
╷
│ Error: unable to make request: request failed, retries exceeded: Get "http://51.142.108.226": context deadline exceeded
│
│   with terracurl_request.test,
│   on main.tf line 61, in resource "terracurl_request" "test":
│   61: resource "terracurl_request" "test" {
│
╵
Releasing state lock. This may take a few moments...
```


looking good :)


and trying with a few other combinations of timeout, retry interval, max retries, etc...

seems to be working as expected

**Comment by devops-rob** [comment]
Thanks @lucymhdavies This is now fixed in #57 I'm cutting a new release now
